aspnet/testing/integration-testing.rst

```diff
@@ -1,8 +1,154 @@
+
+The Test Host
+^^^^^^^^^^^^^
+ASP.NET includes a test host that can be added to integration test projects and used to host ASP.NET applications, serving test requests without the need for  real web host. The provided sample includes an integration test project which has been configured to use `xUnit <https://xunit.github.io/>`_ and the Test Host, as you can see from this excerpt from its ``project.json`` file:
+
+.. literalinclude:: integration-testing/sample/test/PrimeWeb.IntegrationTests/project.json
+  :linenos:
+  :language: javascript
+  :lines: 21-26
+  :dedent: 2
+  :emphasize-lines: 5
+
+Once the Microsoft.AspNet.TestHost package is included in the project, you will be able to create and configure a TestServer in your tests. The following test shows how to verify that a request made to the root of a site returns "Hello World!" and should run successfully against the default ASP.NET Empty Web template created by Visual Studio.
+
+.. literalinclude:: integration-testing/sample/test/PrimeWeb.IntegrationTests/PrimeWebDefaultRequestShould.cs
+  :linenos:
+  :language: c#
+  :lines: 10-30
+  :dedent: 4
+  :emphasize-lines: 7
+
+These tests are using the Arrange-Act-Assert pattern, but in this case all of the Arrange step is done in the constructor, which creates an instance of ``TestServer``. There are several different ways to configure a ``TestServer`` when you create it; in this example we are passing in the ``Configure`` method from our system under test (SUT)'s ``Startup`` class. This method will be used to configure the request pipeline of the ``TestServer`` identically to how the SUT server would be configured.
+
+In the Act portion of the test, a request is made to the ``TestServer`` instance for the "/" path, and the response is read back into a string. This string is then compared with the expected string of "Hello World!". If they match, the test passes, otherwise it fails.
+
+Now we can add a few additional integration tests to confirm that the prime checking functionality works via the web application:
+
+.. literalinclude:: integration-testing/sample/test/PrimeWeb.IntegrationTests/PrimeWebCheckPrimeShould.cs
+  :linenos:
+  :language: c#
+  :lines: 10-62
+  :dedent: 4
+  :emphasize-lines: 7
+
+Note that we're not really trying to test the correctness of our prime number checker with these tests, but rather that the web application is doing what we expect. We already have unit test coverage that gives us confidence in ``PrimeService``, as you can see here:
+
+.. image:: integration-testing/_static/test-explorer.png
+
+.. note:: You can learn more about the unit tests in the :doc:`unit-testing` article.
+
+Now that we have a set of passing tests, it's a good time to think about whether we're happy with the current way in which we've designed our application. If we see any `code smells <http://deviq.com/code-smells/>`_, now may be a good time to refactor the application to improve its design.
+
+Refactoring
```

Comment:
@danroth Refactoring and testing tend to go hand-in-hand. The general topic of refactoring is probably out of scope for ASP.NET, but this section in particular demonstrates how to refactor ASP.NET to introduce a piece of middleware, a scenario I expect will be somewhat common and therefore appropriate to include. It may make sense to update the section header, though, to "Refactoring to Use Middleware". Thoughts?